src/Infrastructure/LeanCode.AI/CQRSAIFunction.cs

```diff
@@ -0,0 +1,209 @@
+using System.Diagnostics;
+using System.Reflection;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using System.Text.Json.Schema;
+using LeanCode.AI.Contracts;
+using LeanCode.Contracts;
+using LeanCode.CQRS.RemoteHttp.Client;
+using Microsoft.Extensions.AI;
+using Microsoft.Extensions.DependencyInjection;
+using static LeanCode.AI.Contracts.McpToolHints;
+using IOException = System.InvalidOperationException;
+
+namespace LeanCode.AI;
+
+public abstract class CQRSAIFunction(Type cqrsObjectType) : AIFunction
+{
+    public enum JsonSchemaSource
+    {
+        JsonSchemaExporter = 0,
+        AIJsonUtilities = 1,
+    }
+
+    public static readonly McpToolAttribute EmptyMetadata = new(string.Empty);
+
+    public Type CQRSObjectType { get; } = cqrsObjectType;
+    public McpToolAttribute? Metadata { get; } = cqrsObjectType.GetCustomAttribute<McpToolAttribute>();
+    public string? Title { get; protected init; }
+    public McpToolHints ToolHints { get; protected init; }
+
+    protected abstract McpToolHints DefaultHints { get; }
+}
+
+public abstract class CQRSAIFunction<TExecutor, TObject, TResult> : CQRSAIFunction
```

Comment:
Do we need to know the `TObject` type here? Can't we work on `object`s? We can't really do anything with the type either way.

Reply:
We need it to somehow call `GetAsync`, so we'd need this type to construct appropriate generic method one way or another.